versioned/mongodb/src/main/java/com/dremio/nessie/versioned/store/mongodb/BsonConditionVisitor.java

```diff
@@ -0,0 +1,127 @@
+   */
+  private static class BsonValueVisitor implements ValueVisitor<String> {
+    @Override
+    public String visit(Value value) {
+      return toMongoExpr(value.getValue());
+    }
+
+    @Override
+    public String visit(ExpressionFunction value) {
+      final ExpressionFunction.FunctionName name = value.getName();
+      final List<Value> arguments = value.getArguments();
+      if (arguments.size() != name.getArgCount()) {
+        throw new InvalidParameterException(
+          String.format("Number of arguments provided [%d] does not match the number expected [%d] for %s.",
+            arguments.size(), name.getArgCount(), name));
+      }
+
+      switch (name) {
+        case EQUALS:
+          // Special case SIZE, as the object representation is awkward for how it should be represented by MongoDB.
+          if (isSize(arguments.get(0))) {
+            return String.format("{%s: {\"$size\": %s}}",
+              arguments.get(0).getFunction().getArguments().get(0).accept(this), arguments.get(1).accept(this));
+          }
+
+          return String.format("{%s: %s}", arguments.get(0).accept(this), arguments.get(1).accept(this));
+        default:
+          throw new UnsupportedOperationException(String.format("%s is not a supported top-level MongoDB function.", name));
+      }
+    }
+
+    @Override
+    public String visit(ExpressionPath value) {
+      return value.getRoot().accept(BsonPathVisitor.INSTANCE, true);
+    }
+
+    private boolean isSize(Value value) {
+      return (value.getType() == Value.Type.FUNCTION) && (((ExpressionFunction)value).getName() == ExpressionFunction.FunctionName.SIZE);
+    }
+  }
+
+  static final BsonValueVisitor VALUE_VISITOR = new BsonValueVisitor();
+
+  /**
+  * This is a callback method that ConditionExpression will call when this visitor is accepted.
+  * It creates a BSON representation of the ConditionExpression object.
+  * @param conditionExpression to be converted into BSON.
+  * @return the converted ConditionExpression object in BSON format.
+  */
+  @Override
+  public Bson visit(final ConditionExpression conditionExpression) {
+    final String functions = conditionExpression.getFunctions().stream()
+        .map(f -> f.accept(VALUE_VISITOR))
+        .collect(Collectors.joining(", "));
+
+    // This intentionally builds up a JSON representation of the query, and then parses it back to a BSON Object
+    // representation. The reason for this is that the Value hierarchy in Nessie is a combination of Bson and BsonValue
+    // in MongoDB parlance, and those objects don't share a common parent, which makes the standard visitor need to
+    // return both.
+    return BsonDocument.parse(String.format("{\"$and\": [%s]}", functions));
+  }
+
+  /**
+   * Convert the Entity chain to an equivalent MongoDB expression.
+   * @param value the entity to convert.
+   * @return the MongoDB string representation of the Entity.
+   */
+  static String toMongoExpr(Entity value) {
```

Comment:
toMongoRepresentation? or toMongoEntity? (Unless mongo representation are called expressions in Mongo).